### PR TITLE
Three month03 proposal: conjugate gradient method

### PR DESCRIPTION
**Pull request by ThreeMonth03**
---
A good PR should include the following parts:

- [x] A clear title
- [x] A meaningful message for PR, as well as its commits
- [x] from your specific branch (***not master***) merging to teacher's `HW` branch (***not master***, again)
- [x] excluding any irrelevant files, such as binaries, text files, or dot files
- [x] passing tests/CI

### CODE REVIEW

File: project/ThreeMonth03/README.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
+# NSD Proposal
```
Comment (yungyuc) [inline project/ThreeMonth03/README.md:1]: What is the name of your project?

File: project/ThreeMonth03/README.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+
+## Problem to Solve
+
+I want to solve the global minima of the quadratic function by searching the numerical solution. One of the algorithm that can solve the problem is conjugate gradient method.
+
+## Conjugate gradient method
+### Introduction
+
+Conjugate gradient method is a gradient-descent-based method that can find the minima of the quadratic function in the hyper-dimensional space.
```
Comment (yungyuc) [inline project/ThreeMonth03/README.md:17]: What if the function is not quadratic?

File: project/ThreeMonth03/README.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+## Basic Information
+
+I will implement conjugate gradient method as my final project.
+
+Project URL: [https://github.com/ThreeMonth03/Conjugate-Gradient-Method-library](https://github.com/ThreeMonth03/Conjugate-Gradient-Method-library)
+
+
+## Problem to Solve
+
+I want to solve the global minima of the quadratic function by searching the numerical solution. One of the algorithm that can solve the problem is conjugate gradient method.
```
Comment (yungyuc) [inline project/ThreeMonth03/README.md:12]: Please change to use 79-char lines.  Long lines are hard to review.  See #89 as an example for how to use short lines.

File: project/ThreeMonth03/README.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+Conjugate gradient method is pretty suitable for solving the regression problem. The algorithm can help the researchers working in deep learning, optimization and social science. 
+
+## System Architecture
+
+The above algorithm will be encapsulated by c++, and the algorithm also could be called by python with pybind11.
+
+### Constraints
+
+The system assumes that the input $H$ is positive-definite matrix and well-conditioned. 
+The quadratic function exists a global minima if and only if the matrix $H$ is positive-definite. Additionally, the convergence speed is pretty fast only when $H$ is well-conditioned.
```
Comment (yungyuc) [inline project/ThreeMonth03/README.md:67]: There will be little for discussions in your presentation if everything is positive-definite.  (Solution of positive-definite extrema usually does not need a numerical method.)

To make a decent presentation, you probably will need to handle some ill-behaviors.  There should be plenty of literature discussing it.  Pick simple ones to proceed.

File: project/ThreeMonth03/README.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+The quadratic function exists a global minima if and only if the matrix $H$ is positive-definite. Additionally, the convergence speed is pretty fast only when $H$ is well-conditioned. 
+In other words, people should check the constrain by themselves or numpy before using the project.
+
+## API Description
+The estimated API is below:
+### c++
+```c++
+class ConjugateGradient: Algorithm{
+    public:
+        vector<double> const & H(vector<float> const & H) const;
```
Comment (yungyuc) [inline project/ThreeMonth03/README.md:76]: Input 32-bit real number but output higher precision of 64-bits?  It does look like reasonable.  You may need to articulate why.

File: project/ThreeMonth03/README.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+Week 1 (10/31): Implement the basic c++ API without multithread.
+
+Week 2 (11/7):  Finish the basic c++ API without multithreading.
+
+Week 3 (11/14): Wrapping and test for pytest.
+
+Week 4 (11/21): Research how to dot product the matrix fastly.
+
+Week 5 (11/28): Implement the matrix multiplication with multithread.
```
Comment (yungyuc) [inline project/ThreeMonth03/README.md:111]: Parallel computing for optimization is complex enough to take more than one week.

**Review by yungyuc — CHANGES_REQUESTED** [review]
Please adjust.

File: project/ThreeMonth03/README.md (hunk 2 above)
Comment (ThreeMonth03) [reply project/ThreeMonth03/README.md:17]: Conjugate gradient method cannot deal with non-quadratic function.

File: project/ThreeMonth03/README.md (hunk 4 above)
Comment (ThreeMonth03) [reply project/ThreeMonth03/README.md:67]: By the way, I ignore a strict condition of $H$, that $H$ should also be a symmetric matrix.
To be frank, it seems that the application of conjugate gradient method is few. 
Do you suggest that I should find another topic?

File: project/ThreeMonth03/README.md (hunk 5 above)
Comment (ThreeMonth03) [reply project/ThreeMonth03/README.md:76]: Sorry, there are some typos. Thanks you for reminding.

File: project/ThreeMonth03/README.md (hunk 1 above)
Comment (ThreeMonth03) [reply project/ThreeMonth03/README.md:1]: Thank you for reminding, and I will fix it later.

File: project/ThreeMonth03/README.md (hunk 6 above)
Comment (ThreeMonth03) [reply project/ThreeMonth03/README.md:111]: Thank you for reminding, I will change the schedule.

File: project/ThreeMonth03/README.md (hunk 2 above)
Comment (yungyuc) [reply project/ThreeMonth03/README.md:17]: Let me put it this way: What will happen when you use CG to find the minimum of a field that is not quadratic?

File: project/ThreeMonth03/README.md (hunk 4 above)
Comment (yungyuc) [reply project/ThreeMonth03/README.md:67]: I did not suggest you to change a numerical method.  The numerical method is not the point.  The point is how to design and implement the software system surrounding the numerical method.  All numerical methods have limitation.  The software system is to make them useful to the most extent.

File: project/ThreeMonth03/README.md (hunk 5 above)
Comment (yungyuc) [reply project/ThreeMonth03/README.md:76]: You need not to really apologize.  Rhetorical sorry is fine though.

File: project/ThreeMonth03/README.md (hunk 2 above)
Comment (ThreeMonth03) [reply project/ThreeMonth03/README.md:17]: After browsing the article on the net, 
I try to answer the reason why conjugate gradient method cannot solve non-quadratic function.

In fact, the original problem what CG method want to solve is $Ax = b$.
When $A$ is symmetric and positive-definite,
solving $Ax = b$ is equal to find the global minima of $f(x) = \frac{1}{2}x^{T}Hx + b^{T}x + c$, 
according to differentiating the quadratic function.

Thus, if the function is not quadratic, then the differential of the function is not $Ax = b$,
thus the CG method will be failed.

I hope the answer is helpful for you.

Reference: https://blog.csdn.net/Yemiekai/article/details/123999290

Comment (ThreeMonth03) [reply project/ThreeMonth03/README.md:17]: After further reading,
I realize that I neglect the nonlinear conjugate gradient algorithm that solve non-linear target. 
Thus, if the input is non-quadratic funciton and there exists local minima in the function, 
it seems that nonlinear conjugate gradient algorithm can find the local minima.

Therefore, in order to avoid from the troublesome input format and gradient calculation, 
I decide to implement linear conjugate gradient algorithm, 
which can only find the global minimum of quadratic function according to the aforementioned reason.

The reference of non-linear conjugate function: https://indrag49.github.io/Numerical-Optimization/conjugate-gradient-methods-1.html#nonlinear-conjugate-gradient-algorithm

Comment (yungyuc) [reply project/ThreeMonth03/README.md:17]: Let me repeat my question because you did not answer it: What will happen when you use CG to find the minimum of a field that is not quadratic?

Comment (ThreeMonth03) [reply project/ThreeMonth03/README.md:17]: Sorry for my poor English. I guess I answer the question this time. 

Before answering the question, I think I should provide additional information on CG method.
The CG method includes linear CG and non-linear CG algorithm, the primary difference of these methods are the calculations of the step lengths in every iteration.
Linear CG algorithm can precisely calculate the step length in every iteration, but in order to calculate the precise step length, the objective function could only be quadratic function.
Non-linear CG algorithm can approximately calculate the step length by line search, or gradient descent method, and the advantage of non-linear CG algorithm is that the target function could be convex nonlinear objective functions.

Therefore, if non-linear CG algorithm find the minimum $X$ of a field that is not quadratic, then the next iteration of the $X$ is also the minima because the gradient direction is zero vector. Everything is fine.
However, if we attempt to find the non-quadratic function by linear CG algorithm, because the linear CG algorithm cannot calculate the precise step length in every iteration, linear CG method will be failed. The reason why CG algorithm cannot calculate the precise step for non-quadratic function is that the optimization of the step length for non-quadratic function is not quite easy.

In conclusion, I think I should replace all the CG method with linear CG method in README.md, and modify the line 17 "Conjugate gradient method is a gradient-descent-based method" to "Linear conjugate gradient method is a  kind of steepest descent approach". I realize that I confuse gradient descent with steepest descent.

Thank you for your question.

Comment (yungyuc) [reply project/ThreeMonth03/README.md:17]: I think you are moving forward the idea in a right track.  As we discussed today, the key is that you need to consider the input data that are not a polynomial function.  How does your software systems handle it?

Comment (ThreeMonth03) [reply project/ThreeMonth03/README.md:17]: If the input data that is not a polynomial function, the input should be a lot of the points of y and gradients. Thus, in theory, the non-linear conjugate gradient method can deal with the non-quadratic convex input function.

However, I think  some issues need considering when implementing the input data that is not a polynomial function. First, I think the user should input the data point and data gradient in every epoch by themselves, because the non-polynomial inputs, such as voxel and point cloud, are digital such that they are not precise enough. Additionally, the determination of the convex is a trouble.

In conclusion, I think I may deal with the non-quadratic convex input function to a certain degree. It may look like those in below. But I still cannot guarantee the probability of the convergence.

```c++
class CG_Algorithm{
     private:
           bool _is_poly_fucn;
           int epoch;
     public:
           double iteration(const vector<double> & point_xi, const vector<double> & gradient_xi){
                      //......
                      return point_x_i_plus_one;
           }
}
```

Comment (yungyuc) [reply project/ThreeMonth03/README.md:17]: A related question is in what format the input data should be.  After considering the question and summarizing what we have discussed, the proposal may be complete enough for development.

Keep in mind that a complete proposal (or plan) does not guarantee a successful project.  There are a lot of things to do in your execution.

**Comment by yungyuc** [comment]
Please leave a global comment when you want me to review again.  Before requesting, make sure your branch doesn't conflict with `project`, and also comment inline at where you changed and wanted the review.

File: project/README.rst [hunk 7]
```diff
@@ -65,9 +65,13 @@ Follow the format to add your project in lexicographical order:
...
+
 2. `CodeStone1125 <https://github.com/CodeStone1125>`__ for
    `Project subject <CodeStone1125/README.md>`__:
    https://github.com/CodeStone1125/QuadraCompress
 3. `Yu-Rou-Weng <https://github.com/Yu-Rou-Weng>`__ for
    `Project subject <Yu-Rou-Weng/README.md>`__:
    https://github.com/Yu-Rou-Weng/Mesh-Data-Converter
+4. `ThreeMonth03 <https://github.com/ThreeMonth03>`__ for
+   `Project subject <ThreeMonth03/README.md>`__:
+   https://github.com/ThreeMonth03/Conjugate-Gradient-Method-library
```
Comment (ThreeMonth03) [inline project/README.rst:77]: I think I solve the conflict problem this time.

**Comment by ThreeMonth03** [comment]
> LGTM. Please let me know when you want it to be merged.

Thank you for your review and suggestion. I want it to be merged now.
